FlowCrypt/src/main/java/com/flowcrypt/email/ui/activity/fragment/MessageDetailsFragment.kt

```diff
@@ -1729,6 +1763,42 @@ class MessageDetailsFragment : BaseFragment<FragmentMessageDetailsBinding>(), Pr
     }
   }
 
+  private fun bindToAttachmentDownloadManagerService() {
+    Intent(requireContext(), AttachmentDownloadManagerService::class.java).also { intent ->
+      context?.bindService(intent, downloadAttachmentsServiceConnection, Context.BIND_AUTO_CREATE)
+    }
+  }
+
+  private fun handleLoadingProgress(
+    mapWithLatestProgress: Map<String, AttachmentDownloadManagerService.DownloadProgress>
+  ) {
+    val uniqueStringIdSetToBeUpdated = mutableSetOf<String>()
+    val existingMap = attachmentsRecyclerViewAdapter.progressMap
+    uniqueStringIdSetToBeUpdated.addAll(
+      mapWithLatestProgress.keys.toMutableSet().apply { removeAll(existingMap.keys) }
+    )
+    existingMap.forEach { (attachmentId, existingProgress) ->
+      val newProgress = mapWithLatestProgress[attachmentId]
+      if (newProgress == null || existingProgress != newProgress) {
+        uniqueStringIdSetToBeUpdated.add(attachmentId)
+      }
+    }
+
+    attachmentsRecyclerViewAdapter.progressMap.clear()
+    attachmentsRecyclerViewAdapter.progressMap.putAll(mapWithLatestProgress)
+    val currentList = attachmentsRecyclerViewAdapter.currentList
+    uniqueStringIdSetToBeUpdated.forEach { uniqueStringId ->
+      currentList.firstOrNull {
+        it.uniqueStringId == uniqueStringId
+      }?.let { attachmentInfo ->
+        val position = currentList.indexOf(attachmentInfo)
+        if (position != -1) {
```

Comment:
I think we won't need this `if` statement as we are trying to index of an element that we already found in the list.